src/Diagnostics/Debug/StateCheck.jl

```diff
@@ -0,0 +1,556 @@
+        l = length(slist)
+        if l == 0
+            println("# SC  MPIStateArray labeled \"$lab\" has no named symbols.")
+        else
+            for s in slist
+                if printHead
+                    if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+                        println("# SC Creating state check callback labeled \"$lab\" for symbols")
+                    end
+                    printHead = false
+                end
+                if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+                    println("# SC ", s)
+                end
+            end
+        end
+    end;
+
+    ###
+    # Initialize total calls counter for the call back
+    ###
+    nCbCalls = 0;
+
+    ###
+    # Create holder for most recent stats
+    ###
+    curStats_dict = Dict();
+    curStats_flat = [];
+
+    ######
+    # Create the callback
+    ######
+    cb = EveryXSimulationSteps(ntFreq) do (s = false)
+        # Track which timestep this is
+        nCbCalls = nCbCalls + 1
+        nStep = (nCbCalls - 1) * ntFreq + 1
+        nSStr = @sprintf("%7.7d", nStep - 1)
+
+        ## Free previous curStats_flat if there is one 
+        #  (interesting piece of code - it works, but I am not sure its very elegant!)
+        for i in range(1, length = length(curStats_flat))
+            pop!(curStats_flat)
+        end
+
+
+        ## Print header
+        nprec = min(max(1, prec), 20)
+        if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+            println("# SC +++++++++++ClimateMachine StateCheck call-back start+++++++++++++++++")
+            println("# SC  Step  |   Label    |  Field   |                            Stats                       ")
+        end
+        hVarFmt = "%" * sprintf1("%d", nprec + 8) * "s"
+        minStr = sprintf1(hVarFmt, " min() ")
+        maxStr = sprintf1(hVarFmt, " max() ")
+        aveStr = sprintf1(hVarFmt, " mean() ")
+        stdStr = sprintf1(hVarFmt, " std() ")
+        if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+            println(
+                "# SC =======|============|==========|",
+                minStr,
+                "|",
+                maxStr,
+                "|",
+                aveStr,
+                "|",
+                stdStr,
+                "|",
+            )
+        end
+
+        ## Iterate over the set of MPIStateArrays for this callback
+        for f in fields
+            olabel = f[2]
+            olStr = @sprintf("%12.12s", olabel)
+            mArray = f[1]
+
+            # Get descriptor for MPIStateArray
+
+            V = typeof(mArray).parameters[2]
+
+            ## Iterate over fields in each MPIStateArray
+            #  (use ivar to index individual arrays within the MPIStateArray)
+            ivar = 0
+            statsVal_dict = Dict()
+            for i in 1:length(V.names)
+                for n in flattenednames(fieldtype(V, i), prefix = fieldname(V, i))
+                    ivar = ivar + 1
+                    nStr = @sprintf("%9.9s", n)
```

Comment:
@kpamnany this looks similar to what I've implemented in the heat equation tutorial, [`get_vars_from_stack`](https://github.com/CliMA/ClimateMachine.jl/pull/1029/files#diff-484affb4b21233277ef7c161e4377d59R151). I think that we should unify these, and it would probably be nice to have one that operates on the entire stack as well as per grid point.